TE-817 Enable accessibility auditing

jzoldak opened the pull request:
using PhantomJS and Google Chrome Accessibility Developer Tools
@benpatterson @cgoldberg @cptvitamin

benpatterson commented on bok_choy/web_app_test.py:
```diff
@@ -1,16 +1,23 @@
 from .performance import HarListener, HarCapturer
-from selenium.webdriver.support.events import EventFiringWebDriver
+
+AXS_FILE = 'vendor/axs_testing.js'
```
What's this file for? If I were brand new to bok-choy, where could I get it?

benpatterson replied:
Ohhhhh I see it further down. Sorry. Ignore.

benpatterson commented on bok_choy/web_app_test.py:
```diff
@@ -86,6 +93,63 @@ def unique_id(self):
+        See https://github.com/GoogleChrome/accessibility-developer-tools
+
+        Since this needs to inject JavaScript into the browser page, the only
+        way I know how to do this is if you are using PhantomJS as your browser.
```
nit: "the only known way" to do this.

benpatterson commented on bok_choy/web_app_test.py:
```diff
@@ -86,6 +93,63 @@ def unique_id(self):
+            }
+        """
+        if self.browser.name is not 'phantomjs':
+            raise NotImplementedError('Only supported in PhantomJS')
```
nit: just a bit more detail in the error message. Like "accessibility audit only supported..."

benpatterson commented on tests/test_accessibility.py:
```diff
@@ -0,0 +1,26 @@
+from .pages import VisiblePage
+
+
+class AccessibilityTest(WebAppTest):
```
Maybe also add a test for the browser exception (e.g., you're not using phantomjs). I figure that can be mocked for travis.

jzoldak replied:
Travis runs the tests only under firefox. Thus implicitly the test_axs_audit testcase will assert on the exception. And the successful audit is not actually performed under travis, which is why I put in that instruction on how to run under phantom locally.

You're right that it would be better if all this was explicit. And if travis runs the test under phantom.

I tried running all the tests under phantom and some fail. I feel that fixing them is outside the scope of this PR. I'll look into adding another commit which will run selected (maybe just this one) tests under phantom as well as the entire suite under firefox.

benpatterson commented:
Overall I think it looks good. Couple questions:
* Will sphinx pick up the doc for us? I just want to be sure users have a 'how-to' for accessibility testing
* Should we make a global setting to always audit given pages? Er, assuming it is cheap. 
* How will we report results when we're running multiple tests? Does the ghostdriver log summarize at the end of the test session somehow?

jzoldak commented:
* Yes sphinx will pick up the docs, but only on official releases - which either @wedaly or I historically have done. I plan to make a new release after we get this change in, as I will want to bump up the version that edx-platform is using.

* My vision is not to audit automatically, as the interpretation of accessibility standards can be a bit subjective, and also unfortunately pretty much all pages in the real world have at least one violation on them. :( Instead I was envisioning that acceptance testcases are written using the framework to get the report and then make assertions on that report, as exemplified in the test_axs_audit testcase.  With a little bit of usage, I'm anticipating that we will discover the usefulness of adding in some assertion helper methods, but I'm not sure at this point what they should be.

* No results are logged, instead we will use assertions. See above bullet point.

clytwynec commented:
I'm still a bit confused about how this will play out with assertions.   If we can't audit automatically because the standards are subjective, it seems strange to use assertions rather than storing a report so that it can be reviewed.

jzoldak commented:
@benpatterson -
Having done more exploration on the automatic auditing / logging of results, I can't wrap my head around how this option would work in practice. This doesn't fit into the paradigm of automated tests, where each test discretely needs to either pass or to fail; you don't look at the printout of every testcase. And gathering up and printing out results over a series of pages doesn't make sense, as the results won't be tied to any objective description out of context, like a page URL because they can be dependent on interactions that you've had with the page. E.g. a certain test may have clicked an element that caused other elements to be hidden, classes applied, etc. before running the report.

@clytwynec -
I anticipated that the assertions of each testcase would be written knowing the current audit results at that time, and to either intentionally fail or to catch future regressions. For example, I have a page on which I know that three specific elements fail the rule AX_ARIA_04 (badAriaAttributeValue). I would write a testcase asserting that these three still fail, and no other elements fail, and that no other rules are violated.

benpatterson commented:
@jzoldak Maybe we think about it more like linting than we do testing? That is, we have a target level of compliance (could be 100%) that is cross-tabulated with specific rules. And we have a count of violations?

jzoldak commented:
Incorporated on- and off-line feedback, now includes both an active and passive auditing ability. And documentation. :)
@clytwynec @benpatterson this PR is ready for re-review.

clytwynec commented on docs/accessibility.rst:
```diff
@@ -0,0 +1,182 @@
+
+  * Actively, by calling the Page Object class's do_axs_audit method and then
+  asserting on the results returned.
+  * Passively, by setting the VERIFY_ACCESSIBILITY environment variable to True.
```
This bullet isn't getting formatted properly.  White space issue maybe?

clytwynec commented on docs/accessibility.rst:
```diff
@@ -0,0 +1,182 @@
+--------------------------------------------------
+
+A page object's list of accessibility rules to audit for that page are defined
+in the axs_audit_rules method. The code for the available rules can be found
```
I was confused about what  'code' means here.   The javascript code has code values for each rule in it. (e.g. BadAriaAttributeValue.js has `code: 'AX_ARIA_04'` in the addRule call) 

I was then confused why ''badAriaAttributeValue'' was being used in your example below.    It might be helpful to specify "javascript code" here, and to mention that the important part to look for is the 'name' attribute used in addRule.

clytwynec commented on docs/accessibility.rst:
```diff
@@ -0,0 +1,182 @@
+.. code-block:: python
+
+    def axs_audit_rules(self):
+        return ['badAriaAttributeValue', 'imagesWithoutAltText']
```
Is there a way to say "use all of the rules _except_ for this subset'' ?

jzoldak replied:
The js that we're using does have a way to do this, but I only plumbed in the rules to run, not the rules to ignore. Was planning on leaving that for a future implementation. See [Audit.js](https://github.com/GoogleChrome/accessibility-developer-tools/blob/master/src/js/Audit.js#L48-60)

That said, I'm realizing it'd probably be better to rename this "audit_rules_to_run" right now in this first implementation, to leave space for adding "audit_rules_to_ignore" later.

clytwynec commented on docs/accessibility.rst:
```diff
@@ -0,0 +1,182 @@
+The audit rules that are checked and optionally the scope of the audit
+within the webpage itself are defined on each Page Object's definition.
+
+The general methodology for enabling accessibility auditing is:
```
I just noticed that these bullet points actually outline the sections of the rest of the page.  Can these be linked to the corresponding sections?

jzoldak replied:
omg I can't believe you're making me learn rst.

clytwynec replied:
I _think_ you can just use the same syntax as the other links, but use the heading texts. (  \`My header text\`_  )

jzoldak replied:
ok done.

clytwynec commented on run_tests.sh:
```diff
@@ -39,7 +39,7 @@ python -m tests.http_server $SERVER_PORT 2> /dev/null &
 
 echo_task "Run unit test suite"
-coverage run -m nose $REPO_ROOT/tests/$1
+coverage run -m nose $REPO_ROOT/tests/$1 --nocapture
```
Why is this change needed?

jzoldak replied:
oh snaps, didn't mean to check that in! (It causes print statements to not be swallowed up by nose, so I use it for debugging.)

clytwynec commented on bok_choy/page_object.py:
```diff
@@ -2,21 +2,28 @@
 
 
+AXS_FILE = 'vendor/axs_testing.js'
+AuditResults = namedtuple('AuditResults', 'errors, warnings')
```
Is there a reason to use `'errors, warnings'` over `['errors', 'warnings']` here?

jzoldak replied:
Oh, both formats mean the same time. I just picked one. Unless that's not what you meant?
https://docs.python.org/2/library/collections.html#collections.namedtuple
> The field_names are a sequence of strings such as ['x', 'y']. Alternatively, field_names can be a single string with each fieldname separated by whitespace and/or commas, for example 'x y' or 'x, y'.

clytwynec replied:
The list format just _seems_ more pythonic to me, but I think it's just a personal preference.

clytwynec commented on bok_choy/page_object.py:
```diff
@@ -173,6 +187,7 @@ def __init__(self, browser):
             PageObject
         """
         self.browser = browser
+        self.verify_accessibility = bool(os.environ.get('VERIFY_ACCESSIBILITY'))
```
What happens if I `export VERIFY_ACCESSIBILITY=False` or `export VERIFY_ACCESSIBILITY=0`?   I think that would result in self.verify_accessibility being True.

benpatterson commented on bok_choy/page_object.py:
```diff
@@ -500,3 +520,130 @@ def wait_for_element_invisibility(self, element_selector, description, timeout=6
         """
         self.wait_for(lambda: not self.q(css=element_selector).visible, description=description, timeout=timeout)
+
+    def axs_audit_rules(self):
```
Not a must, but would it make sense to have these in a different class that gets inherited by PageObject?

benpatterson commented on bok_choy/page_object.py:
```diff
@@ -500,3 +520,130 @@ def wait_for_element_invisibility(self, element_selector, description, timeout=6
+        return None
+
+    def axs_scope(self):
+        """
```
This is awesome; so glad you built this in here.

clytwynec replied:
I agree.  Would this allow us to do something like measure the accessibility of a specific xblock in the platform?

jzoldak replied:
Yes. Or test the content of courseware, ignoring the surrounding lms chrome (which will be tested in the context of the platform itself), etc.

benpatterson commented on bok_choy/page_object.py:
```diff
@@ -500,3 +520,130 @@ def wait_for_element_invisibility(self, element_selector, description, timeout=6
+
+            None if the page object has no rules defined to check.
+        """
+        if self.browser.name != 'phantomjs':
```
Should we raise an exception when someone has asked for verify accessibility, but has no rules?

benpatterson commented on bok_choy/page_object.py:
```diff
@@ -500,3 +520,130 @@ def wait_for_element_invisibility(self, element_selector, description, timeout=6
+            raise NotImplementedError(msg)
+
+        rules = self.axs_audit_rules()
+        if rules is None:
```
Sorry this is where I should have asked the question...should we raise an exception when no rules are defined, but someone's asking for an audit?

benpatterson commented on docs/accessibility.rst:
```diff
@@ -0,0 +1,182 @@
+as each of those will call out to wait_for_page().
+
+If any assessibility errors are found, then the testcase will fail with an
+AccessibilityError.
```
Could you give an example here? Just to make it easy-peasy for python novices to understand.

benpatterson commented on vendor/axs_testing.js:
```diff
@@ -0,0 +1,1710 @@
+/*
```
Not a must: should the path be vendor/google/axs_testing.js?

benpatterson commented on tests/test_accessibility.py:
```diff
@@ -0,0 +1,128 @@
+    @patch('tests.pages.AccessibilityPage.axs_scope')
+    @patch('tests.pages.AccessibilityPage.axs_audit_rules')
+    def test_axs_audit_limit_scope(self, mock_rules, mock_scope):
+        # Limit the scope to the div with AX_ARIA_04 violations
```
Nit...make these docstrings?

benpatterson commented:
@jzoldak my big question is whether or not we need to raise an exception if someone asked for implicit testing but has set no rules. Or, perhaps that will come with the default (all) rules?

jzoldak commented:
@benpatterson after talking this through with @clytwynec, changed the implementation to check all rules by default. This will cover the use case "As a developer I turn on implicit rules checking and see tons of of errors as my tests traverse my application."

We think it will be useful, and can later implement, overriding that default via environment variable.

And also I added a warning if rules for a page are checked, either actively or implicitly, and the page object is configured to not check any rules.

benpatterson commented:
@jzoldak just to clarify, that would be "check all rules by default [when I set the env variable]", is that right?

In other words..this sounds great. Assuming that I could still--by contrast--have my own set of rules to use on the tests I specify...

jzoldak commented:
@benpatterson The hierarchy we are currently proposing:
* (TBD, not implemented, still under discussion) takes precedence: appropriately named feature flag saying no matter what I have defined elsewhere override and run with these specified rules
* (currently implemented): defined on page object
* (TBD, not implemented): appropriately named feature flag saying if not specified on the page object, use these rules (overriding the default below)
* (currently implemented): default to using all rules

benpatterson commented:
Great summary. SGTM.

benpatterson commented:
:+1:

cptvitamin commented:
very excited to see progress on this @jzoldak @clytwynec @benpatterson.  Let me know when you have something for me to review.  Looking forward to seeing this in action.

clytwynec commented:
:+1:
